kafka-connector/src/main/java/com/google/pubsub/kafka/source/CloudPubSubSourceTask.java

```diff
@@ -136,14 +134,19 @@ public void start(Map<String, String> props) {
         }
         ackIds.add(ackId);
         Map<String, String> messageAttributes = message.getAttributes();
-        String key = messageAttributes.get(kafkaMessageKeyAttribute);
+
         Long timestamp = getLongValue(messageAttributes.get(kafkaMessageTimestampAttribute));
-        if (timestamp == null){
+        if (timestamp == null) {
           timestamp = Timestamps.toMillis(message.getPublishTime());
         }
         ByteString messageData = message.getData();
         byte[] messageBytes = messageData.toByteArray();
 
+        String key = messageAttributes.get(kafkaMessageKeyAttribute);
+        if (key == null && kafkaPartitionScheme.equals(PartitionScheme.HASH_VALUE)) {
+          key = Integer.toString(Utils.murmur2(messageBytes));
```

Comment:
This is a backwards-incompatible change and we should probably be careful with such a thing. If one is using the connector in any kind of distributed fashion, moving to a new version could result in inconsistent hashing. Can you make this a new partition scheme to maintain backwards compatibility?

Reply:
Well previous version is sort of incompatible with itself since `Object#hashCode` is not the same for the same object in different vms. I can revert it back though if you think fixing it is risky.